### PR TITLE
twister: Add support for gTest as a harness

### PR DESCRIPTION
**Pull request by yperess**
Some 3P logic (Pigweed for example) is already heavily invested in gTest as a testing framework. Adding the `gtest` harness enables running the existing 3P tests on various Zephyr platforms and configs.

### CODE REVIEW

**Comment by hakehuang** [comment]
do we have a example for gtest in side zephyr test/sample? and please update a gtest document in twister document for gtest usage.

**Comment by yperess** [comment]
> do we have a example for gtest in side zephyr test/sample? and please update a gtest document in twister document for gtest usage.

We don't have an example of gTest, currently gTest isn't a Zephyr module so we would first need to get approval from the TSC to create the fork of gTest, then we could add samples.

Regarding the docs, what did you have in mind? I've updated `twister.rst` and from what I can tell the gtest harness has as much documentation as the others. I'm happy to expand the line to something like:
```
- gtest: Use this harness to capture output from unit tests written using the gTest framework.
```

The issue is that it feels odd that the other harnesses don't have a description along with them.

**Comment by hakehuang** [comment]
> > do we have a example for gtest in side zephyr test/sample? and please update a gtest document in twister document for gtest usage.
> 
> We don't have an example of gTest, currently gTest isn't a Zephyr module so we would first need to get approval from the TSC to create the fork of gTest, then we could add samples.
> 
do you have plan to add the gtest in zephyr twister tree? add a plugin for what can't use seems strange.

> Regarding the docs, what did you have in mind? I've updated `twister.rst` and from what I can tell the gtest harness has as much documentation as the others. I'm happy to expand the line to something like:
> 
> ```
> - gtest: Use this harness to capture output from unit tests written using the gTest framework.
> ```
> The issue is that it feels odd that the other harnesses don't have a description along with them.

The old harness need document is necessary if they are kind of framework bases. but most of them are quit simple. if the Gtest framework has much potential, then it is better to have a getting start guide for  community to us. unless you intention is to keep it in limited user case.

**Comment by yperess** [comment]
> > > do we have a example for gtest in side zephyr test/sample? and please update a gtest document in twister document for gtest usage.
> > 
> > 
> > We don't have an example of gTest, currently gTest isn't a Zephyr module so we would first need to get approval from the TSC to create the fork of gTest, then we could add samples.
> 
> do you have plan to add the gtest in zephyr twister tree? add a plugin for what can't use seems strange.

I kinda have plans for it. I'm working on integrating Pigweed (pigweed.dev) with Zephyr. Pigweed already has extensive tests written in gTest. So we needed a way to run all these tests on Zephyr's platforms. This harness enables that. Soon after ZDS we'll be asking to make Pigweed an official module of Zephyr (https://github.com/zephyrproject-rtos/zephyr/issues/53851). When this happens, we'll be using the tests which will be running in Pigweed's CI (to make sure new Pigweed changes don't break against the latest release of Zephyr) but on Zephyr to make sure the ToT changes to Zephyr don't break against the latest Pigweed.
 
> > Regarding the docs, what did you have in mind? I've updated `twister.rst` and from what I can tell the gtest harness has as much documentation as the others. I'm happy to expand the line to something like:
> > ```
> > - gtest: Use this harness to capture output from unit tests written using the gTest framework.
> > ```
> > The issue is that it feels odd that the other harnesses don't have a description along with them.
> 
> The old harness need document is necessary if they are kind of framework bases. but most of them are quit simple. if the Gtest framework has much potential, then it is better to have a getting start guide for community to us. unless you intention is to keep it in limited user case.

I'm happy to document if you have any suggestions, but this is really just "take an existing gTest application, link it to Zephyr's `app` target, and run". There's not much more to it other than "You can now run tests built on gTest".

**Comment by yperess** [comment]
@hakehuang I've added an extra line detailing when to use `gtest` as the harness

File: doc/develop/test/twister.rst [hunk 1]
```diff
@@ -417,6 +417,8 @@ harness: <string>
     - test
     - console
     - pytest
+    - gtest: Use this harness if you've already got tests written in the gTest
+      framework and do not wish to update them to zTest.
```
Comment (PerMac) [inline doc/develop/test/twister.rst:421]: the previous paragraph says: " The console harness tells Twister to parse a test's text output for a regex defined in the test's YAML file.". Maybe instead of having the comment for gtest here, where harnesses are listed, move it to a new paragraph, just after the list. Also move the sentence from before. E.g.

`Harnesses ztest, gtest and console are based on parsing of the output and matching certain phrases. Ztest and gtests look for pass/fail/etc. frames defined in those frameworks. The console harness tells Twister to parse a test's text output for a regex defined in the test's YAML file.` ?

File: scripts/pylib/twister/twisterlib/harness.py [hunk 2]
```diff
@@ -12,7 +12,7 @@
 logger.setLevel(logging.DEBUG)
 
 # pylint: disable=anomalous-backslash-in-string
-result_re = re.compile(".*(PASS|FAIL|SKIP) - (test_)?(.*) in (\d*[.,]?\d*) seconds")
+result_re = re.compile(".*(PASS|FAIL|SKIP) - (test_)?(.*) in (\\d*[.,]?\\d*) seconds")
```
Comment (PerMac) [inline scripts/pylib/twister/twisterlib/harness.py:15]: I don't see this variable to be used anywhere in this PR, why the change?

File: scripts/pylib/twister/twisterlib/testinstance.py [hunk 3]
```diff
@@ -28,6 +28,8 @@ class TestInstance:
         out directory used is <outdir>/<platform>/<test case name>
     """
 
+    __test__ = False
```
Comment (PerMac) [inline scripts/pylib/twister/twisterlib/testinstance.py:31]: why this change?

**Review by PerMac — CHANGES_REQUESTED** [review]
Only minor questions/comments. Otherwise LGTM. I only looked through the code, haven't run any test to see it in action

File: doc/develop/test/twister.rst (hunk 1 above)
Comment (yperess) [reply doc/develop/test/twister.rst:421]: That's way better than anything I've come up with, thanks!

File: scripts/pylib/twister/twisterlib/harness.py (hunk 2 above)
Comment (yperess) [reply scripts/pylib/twister/twisterlib/harness.py:15]: My IDE was complaining that this was incorrect. I'm not sure what this regex was used for, but for \d to work it definitely needed the extra '\' so this is just my own personal guidelines of "leave the code cleaner than you found it" I'm happy to revert if you'd like me to.

Comment (PerMac) [reply scripts/pylib/twister/twisterlib/harness.py:15]: I see. I am fine with such explanation. The CI passes, so I believe it is not breaking its performance

File: scripts/pylib/twister/twisterlib/testinstance.py (hunk 3 above)
Comment (PerMac) [reply scripts/pylib/twister/twisterlib/testinstance.py:31]: Thanks for the explanation

File: doc/develop/test/twister.rst [hunk 4]
```diff
@@ -408,15 +408,22 @@ harness: <string>
...
-    The console harness tells Twister to parse a test's text output for a regex
-    defined in the test's YAML file.
 
     Supported harnesses:
 
     - ztest
     - test
     - console
     - pytest
+    - gtest: Use this harness if you've already got tests written in the gTest
```
Comment (PerMac) [inline doc/develop/test/twister.rst:418]: you forgot to remove the "Use this harness.. " from here

**Review by PerMac — CHANGES_REQUESTED** [review]

File: doc/develop/test/twister.rst (hunk 4 above)
Comment (yperess) [reply doc/develop/test/twister.rst:418]: Oy, sorry, give me 5 minutes.

**Review by PerMac — DISMISSED** [review]

**Review by asemjonovs — DISMISSED** [review]

**Review by alevkoy — DISMISSED** [review]

**Comment by yperess** [comment]
@nashif I'm sure you're swamped this week, but any chance you could take a look at this before the window closes? Sorry for the close call.

**Comment by yperess** [comment]
@PerMac @alevkoy @asemjonovs I had to rebase because of a merge conflict in the release notes. Sorry.

**Review by alevkoy — APPROVED** [review]

**Review by asemjonovs — APPROVED** [review]

**Review by PerMac — APPROVED** [review]

**Review by nashif — APPROVED** [review]
